File: gtci/10-level-order-traversal/python3/6-even-odd-tree.py
from collections import deque


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def isEvenOddTree(self, root: TreeNode) -> bool:
        if root is None:
            return True

        queue = deque()
        queue.append(root)

        level = -1
        while queue:
            level_size = len(queue)
            level += 1

            prev_node_val = None
            for _ in range(level_size):
                node = queue.popleft()

                if level % 2 == 0:
                    if node.val % 2 == 0 or (
                        prev_node_val is not None and node.val <= prev_node_val
                    ):
                        return False
                else:
                    if node.val % 2 != 0 or (
                        prev_node_val is not None and node.val >= prev_node_val
                    ):
                        return False

                if node.left is not None:
                    queue.append(node.left)

                if node.right is not None:
                    queue.append(node.right)

                prev_node_val = node.val

        return True


if __name__ == "__main__":
    sol = Solution()

    root = TreeNode(1)
    root = TreeNode(1)
    root.left = TreeNode(10)
    root.right = TreeNode(4)
    root.left.left = TreeNode(3)
    root.left.right = TreeNode(7)
    assert sol.isEvenOddTree(root) is True

    root = TreeNode(5)
    root.left = TreeNode(9)
    root.right = TreeNode(3)
    root.left.left = TreeNode(12)
    root.right.right = TreeNode(8)
    assert sol.isEvenOddTree(root) is False

    root = TreeNode(7)
    root.left = TreeNode(10)
    root.right = TreeNode(2)
    root.left.left = TreeNode(12)
    root.left.right = TreeNode(8)
    assert sol.isEvenOddTree(root) is False

    print("All test cases passed.")

File: gtci/10-level-order-traversal/cpp/6-even-odd-tree.cpp
#include <cassert>
#include <climits>
#include <iostream>
#include <queue>

#include "tree_utils.h"

using namespace std;

class Solution {
public:
    bool isEvenOddTree(TreeNode* root) {
        if (root == nullptr)
            return true;
        
        std::queue<TreeNode*> queue;
        queue.push(root);

        int level = -1;
        while (!queue.empty()) {
            const size_t levelSize = queue.size();
            level++;

            int prevNodeValue;
            if (level % 2 == 0)
                prevNodeValue = std::numeric_limits<int>::min();
            else
                prevNodeValue = std::numeric_limits<int>::max();

            for (size_t i = 0; i < levelSize; i++) {
                TreeNode* node = queue.front();
                queue.pop();

                if (level % 2 == 0) {
                    if (node->val % 2 == 0 || node->val <= prevNodeValue)
                        return false;
                } else {
                    if (node->val % 2 != 0 || node->val >= prevNodeValue)
                        return false;
                }

                if (node->left != nullptr)
                    queue.push(node->left);
                
                if (node->right != nullptr)
                    queue.push(node->right);
                
                prevNodeValue = node->val;
            }
        }

        return true;
    }
};

int main() {
    Solution sol = Solution();

    TreeNode* root1 = new TreeNode(1);
    root1->left = new TreeNode(10);
    root1->right = new TreeNode(4);
    root1->left->left = new TreeNode(3);
    root1->left->right = new TreeNode(7);
    assert(sol.isEvenOddTree(root1));

    TreeNode* root2 = new TreeNode(5);
    root2->left = new TreeNode(9);
    root2->right = new TreeNode(3);
    root2->left->left = new TreeNode(12);
    root2->right->right = new TreeNode(8);
    assert(!sol.isEvenOddTree(root2));

    TreeNode* root3 = new TreeNode(7);
    root3->left = new TreeNode(10);
    root3->right = new TreeNode(2);
    root3->left->left = new TreeNode(12);
    root3->left->right = new TreeNode(8);
    assert(!sol.isEvenOddTree(root3));

    std::cout << "All test cases passed." << std::endl;

    cleanupTree(root1);
    cleanupTree(root2);
    cleanupTree(root3);

    return 0;
}
